From 88af54625591c61be2f41e530762b041bcb6a581 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 3 Apr 2023 12:38:31 +0200
Subject: [PATCH 0442/1695] pta: scmi: support Ocall2 on MSG channel

Implements Ocall2 sequence for MSG SCMI channels. An MSG SCMI channel is
an SCMI channel using OP-TEE native shared memory. Agent (Linux)
passes the SCM shm reference when invoking SCMI PTA to create an Ocall2
threaded context (PTA command PTA_SCMI_CMD_OCALL2_MSG_THREAD) instead
of using a statically allocated and mapped IOMEM shred memory buffer.

Ocall2 second parameter #2 is used to pass the SCMI message size in
receive and transmit communication.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git a0537dd111e8eab9b7b4ee3bd08949c73d12c24c pta: scmi: support Ocall2 on MSG channel]
Change-Id: I7afaaf3a3d0329942a8d0d41afc840095b1ce3c1
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/pta/scmi.c                       | 77 +++++++++++++++++++++++++--
 lib/libutee/include/pta_scmi_client.h | 45 ++++++++++++----
 2 files changed, 106 insertions(+), 16 deletions(-)

diff --git a/core/pta/scmi.c b/core/pta/scmi.c
index 134ba971f..300a2f3dd 100644
--- a/core/pta/scmi.c
+++ b/core/pta/scmi.c
@@ -15,6 +15,14 @@
 #include <string.h>
 #include <util.h>
 
+struct msg_shm {
+	void *in_buf;
+	size_t in_size;
+	void *out_buf;
+	size_t out_size;
+	size_t out_max_size;
+};
+
 static bool valid_caps(unsigned int caps)
 {
 	return (caps & ~PTA_SCMI_CAPS_VALID_MASK) == 0;
@@ -174,25 +182,42 @@ static TEE_Result cmd_process_msg_channel(uint32_t ptypes,
 }
 
 /* Process an OCALL RPC to client and report status */
-static enum optee_scmi_ocall_reply pta_scmi_ocall(uint32_t channel_id)
+static enum optee_scmi_ocall_reply pta_scmi_ocall(uint32_t channel_id,
+						  struct msg_shm *dyn_shm)
 {
 	uint32_t ocall2_params[2] = {
 		PTA_SCMI_OCALL_CMD_THREAD_READY,
 	};
 
+	if (dyn_shm)
+		ocall2_params[1] = dyn_shm->out_size;
+
 	if (thread_rpc_ocall2_cmd(ocall2_params)) {
-		FMSG("Error on channel %u"PRIu32, channel_id);
+		DMSG("Error on channel %u"PRIu32, channel_id);
 		return PTA_SCMI_OCALL_ERROR;
 	}
 
 	switch (ocall2_params[0]) {
 	case PTA_SCMI_OCALL_PROCESS_SMT_CHANNEL:
-		FMSG("Posting message on channel %u"PRIu32, channel_id);
-		if (!IS_ENABLED(CFG_SCMI_MSG_SMT))
+		FMSG("Posting SMT message on channel %u"PRIu32, channel_id);
+		if (!IS_ENABLED(CFG_SCMI_MSG_SMT) || dyn_shm)
 			return PTA_SCMI_OCALL_ERROR;
 
 		scmi_smt_threaded_entry(channel_id);
 		break;
+	case PTA_SCMI_OCALL_PROCESS_MSG_CHANNEL:
+		FMSG("Posting MSG message on channel %u"PRIu32, channel_id);
+		if (!IS_ENABLED(CFG_SCMI_MSG_SHM_MSG) || !dyn_shm)
+			return PTA_SCMI_OCALL_ERROR;
+
+		dyn_shm->in_size = ocall2_params[1];
+		dyn_shm->out_size = dyn_shm->out_max_size;
+
+		if (scmi_msg_threaded_entry(channel_id, dyn_shm->in_buf,
+					    dyn_shm->in_size, dyn_shm->out_buf,
+					    &dyn_shm->out_size))
+			return PTA_SCMI_OCALL_ERROR;
+		break;
 	case PTA_SCMI_OCALL_CLOSE_THREAD:
 		FMSG("Closing channel %u"PRIu32, channel_id);
 		break;
@@ -231,7 +256,7 @@ static TEE_Result cmd_scmi_ocall_smt_thread(uint32_t ptypes,
 
 	FMSG("Enter Ocall thread on channel %u"PRIu32, channel_id);
 	while (1) {
-		switch (pta_scmi_ocall(channel_id)) {
+		switch (pta_scmi_ocall(channel_id, NULL)) {
 		case PTA_SCMI_OCALL_PROCESS_SMT_CHANNEL:
 			continue;
 		case PTA_SCMI_OCALL_CLOSE_THREAD:
@@ -242,6 +267,46 @@ static TEE_Result cmd_scmi_ocall_smt_thread(uint32_t ptypes,
 	}
 }
 
+static TEE_Result cmd_scmi_ocall_msg_thread(uint32_t ptypes,
+					    TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						TEE_PARAM_TYPE_MEMREF_INPUT,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+
+	struct msg_shm dyn_shm = {
+		.in_buf = params[1].memref.buffer,
+		.in_size = params[1].memref.size,
+		.out_buf = params[2].memref.buffer,
+		.out_max_size = params[2].memref.size,
+	};
+	unsigned int channel_id = params[0].value.a;
+	struct scmi_msg_channel *channel = NULL;
+
+	if (!IS_ENABLED(CFG_SCMI_MSG_SHM_MSG))
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (ptypes != exp_pt || !dyn_shm.in_buf || !dyn_shm.out_buf)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	channel = plat_scmi_get_channel(channel_id);
+	if (!channel)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	while (1) {
+		switch (pta_scmi_ocall(channel_id, &dyn_shm)) {
+		case PTA_SCMI_OCALL_PROCESS_MSG_CHANNEL:
+			continue;
+		case PTA_SCMI_OCALL_CLOSE_THREAD:
+			params[2].memref.size =  dyn_shm.out_size;
+			return TEE_SUCCESS;
+		default:
+			return TEE_ERROR_GENERIC;
+		}
+	}
+}
+
 static TEE_Result cmd_get_channel_handle(uint32_t ptypes,
 					 TEE_Param params[TEE_NUM_PARAMS])
 {
@@ -317,6 +382,8 @@ static TEE_Result pta_scmi_invoke_command(void *session __unused, uint32_t cmd,
 		return cmd_get_channel_handle(ptypes, params);
 	case PTA_SCMI_CMD_OCALL2_SMT_THREAD:
 		return cmd_scmi_ocall_smt_thread(ptypes, params);
+	case PTA_SCMI_CMD_OCALL2_MSG_THREAD:
+		return cmd_scmi_ocall_msg_thread(ptypes, params);
 	default:
 		return TEE_ERROR_NOT_SUPPORTED;
 	}
diff --git a/lib/libutee/include/pta_scmi_client.h b/lib/libutee/include/pta_scmi_client.h
index 82ae73c6d..629a5a831 100644
--- a/lib/libutee/include/pta_scmi_client.h
+++ b/lib/libutee/include/pta_scmi_client.h
@@ -69,16 +69,29 @@
 #define PTA_SCMI_CMD_PROCESS_MSG_CHANNEL	4
 
 /*
- * PTA_SCMI_CMD_OCALL2_SMT_THREAD - Allocate a threaded path using Ocall2
+ * PTA_SCMI_CMD_OCALL2_SMT_THREAD - Allocate a thread context using Ocall2
+ * for processing of SMT messages.
  *
  * [in]     value[0].a: channel handle
  *
  * Use Ocall2 support to create a provisioned OP-TEE thread context for
- * the channel. Successful creation of the thread makes this command to
- * return with Ocall command PTA_SCMI_OCALL_CMD_THREAD_READY.
+ * an SMT based SCMI channel. Successful creation of the thread makes this
+ * command to return with Ocall command PTA_SCMI_OCALL_CMD_THREAD_READY.
  */
 #define PTA_SCMI_CMD_OCALL2_SMT_THREAD		2048
 
+/*
+ * PTA_SCMI_CMD_OCALL2_MSG_THREAD - Allocate an thread context using Ocall2
+ * for processing of MSG messages.
+ *
+ * [in]     value[0].a: channel handle
+ *
+ * Use Ocall2 support to create a provisioned OP-TEE thread context for
+ * an MSG based SCMI channel. Successful creation of the thread makes this
+ * command to return with Ocall command PTA_SCMI_OCALL_CMD_THREAD_READY.
+ */
+#define PTA_SCMI_CMD_OCALL2_MSG_THREAD		2049
+
 /*
  * Capabilities
  */
@@ -108,24 +121,33 @@
  *
  * At channel setup, we start from the REE: caller requests an Ocall context.
  *
- * 0. REE opens a session toward PTA SCMI. REE invokes PTA command
- *    PTA_SCMI_CMD_GET_CHANNEL_HANDLE to get a channel handler.
+ * 0. REE client opens a session toward PTA SCMI. Client invokes PTA command
+ *    PTA_SCMI_CMD_GET_CHANNEL_HANDLE to get a channel handler and passes
+ *    capabilities for querying which of SMT or MSG shared memory protocol is
+ *    to be used and whether OCall can be used to suspend SCMI task until next
+ *    incoming SCMI message.
  *
- * 1. REE invokes command PTA_SCMI_CMD_OCALL2_SMT_THREAD with an Ocall context.
+ * 1. When client and OP-TEE core both support Ocall context management,
+ *    REE invokes command PTA_SCMI_CMD_OCALL2_SMT_THREAD (or respectively
+ *    PTA_SCMI_CMD_OCALL2_MSG_THREAD) with an Ocall context.
  *    This is the initial invocation of the Ocall thread context. Any further
  *    error in the thread communication make the Ocall to return from REE to
  *    TEE with an error status (Ocall2 out_param1 == 0) upon which SCMI PTA
- *    will return from initial command PTA_SCMI_CMD_OCALL2_SMT_THREAD with an
- *    error result.
+ *    will return from initial command PTA_SCMI_CMD_OCALL2_SMT_THREAD
+ *    (or resp. PTA_SCMI_CMD_OCALL2_MSG_THREAD) with an error result.
+ *    Whether SMT or MSG shared memory protocol is used depends on the
+ *    capabilities exchanged at step 0.
  *
  * 2. Upon support of Ocall the PTA creates an Ocall context and returns to
  *    REE with Ocall command PTA_SCMI_OCALL_CMD_THREAD_READY.
  *
- * 3. REE returns to the PTA, from the Ocall, with output out_param1
- *    set to PTA_SCMI_OCALL_PROCESS_SMT_CHANNEL to post an SCMI message.
+ * 3. When an SCMI message is posted, REE returns to the PTA, from the Ocall,
+ *    with output out_param1 set to PTA_SCMI_OCALL_PROCESS_SMT_CHANNEL
+ *    (or respectively PTA_SCMI_CMD_OCALL2_MSG_THREAD) to post an SCMI message.
  *    In such case, OP-TEE processes the message and returns to REE with
  *    Ocall command PTA_SCMI_OCALL_CMD_THREAD_READY. The SCMI response is in
- *    the shared memory buffer.
+ *    the shared memory buffer. Whether SMT or MSG shared memory protocol is
+ *    used depends on the capabilities exchanged at step 0.
  *
  * 4. Alternatively REE can return from the Ocall with out_param1 set to
  *    PTA_SCMI_OCALL_CLOSE_THREAD. This requests OP-TEE to terminate the
@@ -150,5 +172,6 @@ enum optee_scmi_ocall_reply {
 	PTA_SCMI_OCALL_ERROR = 0, /* OPTEE_RPC_OCALL2_OUT_PARAM1_ERROR */
 	PTA_SCMI_OCALL_CLOSE_THREAD = 1,
 	PTA_SCMI_OCALL_PROCESS_SMT_CHANNEL = 2,
+	PTA_SCMI_OCALL_PROCESS_MSG_CHANNEL = 3,
 };
 #endif /* SCMI_PTA_SCMI_CLIENT_H */
-- 
2.47.2

